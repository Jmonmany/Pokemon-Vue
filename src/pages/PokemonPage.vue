<template>
  <h1 v-if="!selectedPokemon">Wait a bit please...</h1>
  <div v-else>
    <h1>Who is this Pokemon?</h1>
    <Picture :pokemonId="selectedPokemon.id" :showPokemon="showSwitch" />
    <!-- this component will receive the pokemonId and showPokemon props. It will be responsible for displaying the pokemon shadow image and
  only reveal the pokemon image when showPokemon is true. -->
    <Options :pokeOptions="pokemonArray" @selection="checkAnswer($event)" />
    <!-- this component will be responsible for displaying the options for the user to choose from. To do so needs getOptions helper-->
    <!-- selection method is a custom method we create on Options.vue -->
    <div v-if="showAnswer">
      <h2 class="fade-in">{{ message }}</h2>
      <button @click="newGame()">New Game</button>
    </div>
    <!-- we can substitute <div v-if> tag with template tag, to avoid unnecessary divs in the DOM -->
  </div>
</template>

<script>
import Options from "@/components/Options.vue";
import Picture from "@/components/Picture.vue";
import getPokemonOptions from "@/helpers/getOptions";

export default {
  components: {
    Options,
    Picture,
  },
  data() {
    return {
      pokemonArray: [],
      selectedPokemon: null,
      showSwitch: false,
      showAnswer: false,
      message: "",
    };
  },
  methods: {
    async getMixedOptions() {
      // call the getPokemonOptions helper function and assign the result to pokemonArray
      this.pokemonArray = await getPokemonOptions();
      // if we print pokemonArray here, we will see that it is an reactive object.
      // this means that if we change pokemonArray, the component will be re-rendered
      const randomNumber = Math.floor(Math.random() * 4);
      this.selectedPokemon = this.pokemonArray[randomNumber];
    },
    checkAnswer(id) {
      this.showSwitch = true;
      this.showAnswer = true;
      if (id === this.selectedPokemon.id) {
        this.message = `Correct! The pokemon is ${this.selectedPokemon.name}`;
      } else {
        this.message = `Oops! It was ${this.selectedPokemon.name}`;
      }
    },
    newGame() {
      this.showSwitch = false;
      this.showAnswer = false;
      this.selectedPokemon = null;
      this.getMixedOptions();
    },
  },
  mounted() {
    // mounted is a Vue hook that is called when the component is mounted to the DOM
    this.getMixedOptions();
  },
};
</script>

<style></style>
