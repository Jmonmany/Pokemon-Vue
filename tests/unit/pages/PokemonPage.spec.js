import { shallowMount, mount } from "@vue/test-utils";
import PokemonPage from "@/pages/PokemonPage";
import { pokemonsMock } from "../mocks/pokemons.mock";

describe("PokemonPage.vue component", () => {
  let mixSpy;
  beforeEach(() => {
    mixSpy = jest.spyOn(PokemonPage.methods, "getMixedOptions");
  });
  test("should call the getMixedoptions method when mounted", () => {
    expect(mixSpy).toHaveBeenCalled();
  });
  test("should match the snapshot AFTER render component with some pokemons", () => {
    const wrapper = mount(PokemonPage, {
      data() {
        return {
          pokemonArray: pokemonsMock,
          selectedPokemon: pokemonsMock[0],
          showSwitch: false,
          showAnswer: false,
          message: "",
        };
      },
    });
    expect(wrapper.html()).toMatchSnapshot();
  });
  test("should render Options and Picture components", () => {
    const wrapper = shallowMount(PokemonPage, {
      data() {
        return {
          pokemonArray: pokemonsMock,
          selectedPokemon: pokemonsMock[0],
          showSwitch: false,
          showAnswer: false,
          message: "",
        };
      },
    });
    // console.log(wrapper.vm.$data);
    // this will console de data object
    const options = wrapper.find("options-stub");
    const picture = wrapper.find("picture-stub");
    expect(options.exists()).toBeTruthy();
    expect(picture.exists()).toBeTruthy();
    // console.log(picture.attributes())
    expect(picture.attributes("pokemonid")).toBe("1");
    // apparently the attribute we are sending to the component is pokemonId, but
    // once we console the attributes, it is pokemonid (lowercase i)
    expect(options.attributes("pokeoptions")).toBeTruthy();
    // same here, we are sending pokeOptions, but it is rendered as pokeoptions
  });
  test("checkAnswer method should work correctly", async () => {
    const wrapper = shallowMount(PokemonPage, {
      data() {
        return {
          pokemonArray: pokemonsMock,
          selectedPokemon: pokemonsMock[0],
          showSwitch: false,
          showAnswer: false,
          message: "",
        };
      },
    });
    await wrapper.vm.checkAnswer(1);
    const h2 = wrapper.find("h2");
    expect(h2.exists()).toBeTruthy();
    expect(wrapper.vm.showSwitch).toBeTruthy();
  });
});
