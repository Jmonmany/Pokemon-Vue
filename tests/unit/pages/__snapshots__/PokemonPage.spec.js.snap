// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PokemonPage.vue component should match the snapshot AFTER render component with some pokemons 1`] = `
<div>
  <h1>Who is this Pokemon?</h1>
  <div class="pokemon-container"><img src="https://raw.githubusercontent.com/PokeAPI/sprites/master/sprites/pokemon/other/dream-world/1.svg" alt="pokemon" class="hidden">
    <!--v-if-->
  </div>
  <!-- this component will receive the pokemonId and showPokemon props. It will be responsible for displaying the pokemon shadow image and
  only reveal the pokemon image when showPokemon is true. -->
  <div class="options-container">
    <ul>
      <li>bulbasaur</li>
      <li>ivysaur</li>
      <li>venusaur</li>
      <li>charmander</li><!-- the $emit method is used to emit an event to the parent component -->
    </ul>
  </div><!-- this component will be responsible for displaying the options for the user to choose from. To do so needs getOptions helper-->
  <!-- selection method is a custom method we create on Options.vue -->
  <!--v-if-->
  <!-- we can substitute <div v-if> tag with template tag, to avoid unnecessary divs in the DOM -->
</div>
`;
